libfeed/feed-knowledge-app-proxy.c

```diff
@@ -0,0 +1,256 @@
+  g_clear_pointer (&priv->knowledge_search_object_path, g_free);
+  g_clear_pointer (&priv->knowledge_app_id, g_free);
+
+  G_OBJECT_CLASS (eos_discovery_feed_knowledge_app_proxy_parent_class)->finalize (object);
+}
+
+static void
+eos_discovery_feed_knowledge_app_proxy_class_init (EosDiscoveryFeedKnowledgeAppProxyClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = eos_discovery_feed_knowledge_app_proxy_get_property;
+  object_class->set_property = eos_discovery_feed_knowledge_app_proxy_set_property;
+  object_class->finalize = eos_discovery_feed_knowledge_app_proxy_finalize;
+
+  eos_discovery_feed_knowledge_app_proxy_props[PROP_DBUS_PROXY] =
+    g_param_spec_object ("dbus-proxy",
+                         "DBus Proxy",
+                         "The GDBusProxy for this EosDiscoveryFeedKnowledgeAppProxy",
+                         G_TYPE_DBUS_PROXY,
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  eos_discovery_feed_knowledge_app_proxy_props[PROP_DESKTOP_ID] =
+    g_param_spec_string ("desktop-id",
+                         "Desktop ID",
+                         "The Desktop ID corresponding to where this proxy came from",
+                         "",
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  eos_discovery_feed_knowledge_app_proxy_props[PROP_BUS_NAME] =
+    g_param_spec_string ("bus-name",
+                         "Bus Name",
+                         "The DBus Bus name for this proxy",
+                         "",
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  eos_discovery_feed_knowledge_app_proxy_props[PROP_KNOWLEDGE_SEARCH_OBJECT_PATH] =
+    g_param_spec_string ("knowledge-search-object-path",
+                         "Knowledge Search Object Path",
+                         "The object path to access the knowledge-search interface",
+                         "",
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  eos_discovery_feed_knowledge_app_proxy_props[PROP_KNOWLEDGE_APP_ID] =
+    g_param_spec_string ("knowledge-app-id",
+                         "Knowledge App ID",
+                         "The object path to access the knowledge-search interface",
+                         "",
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  g_object_class_install_properties (object_class,
+                                     NPROPS,
+                                     eos_discovery_feed_knowledge_app_proxy_props);
+}
+
+/**
+ * eos_discovery_feed_knowledge_app_proxy_new:
+ * @dbus_proxy: A #GDBusProxy connected to the provider
+ * @desktop_id: The desktop-id uniquely identifying the source of this provider
+ * @bus_name: Which bus name the app runs on
+ * @knowledge_search_object_path: (nullable): The object path for the knowledge-search interface
+ * @knowledge_app_id: (nullable): The App ID for the corresponding knowledge-app, if any
```

Comment:
The desktop_id is going to contain `.desktop` whereas the knowledge_app_id is the app ID itself (used for accessing the knowledge app search interface). The reason why I kept them separate is that it might be possible to have app app that you launch that isn't a knowledge app (for instance, gnome-software)

Reply:
How does it being a knowledge app or not make a difference?

Reply:
The desktop-id does not always match the app ID. That might be true in the case of a flatpak, but not true for everything else.

That said, I did some extra checks and at least for gnome-software and flatpaks, it is true, so I'd be willing to change this. But perhaps in a follow up commit (I'd rather keep the port as clean as possible so that we do not lose track).